contracts/contracts/strategies/ConvexGeneralizedMetaStrategy.sol

```diff
@@ -0,0 +1,153 @@
+         * balanced either in OUSD direction or 3Crv.
+         *
+         * Analysis has confirmed that: `It is more cost effective to remove liquidity in balanced manner and
+         * make up for the difference with additional swap. Comparing to removing liquidity in imbalanced manner.`
+         * Results of analysis here: https://docs.google.com/spreadsheets/d/1DYSyYwHqxRzSJh9dYkY5kcgP_K5gku6N2mQVhoH33vY
+         * run it yourself using code in brownie/scripts/liqidity_test.py
+         */
+        // slither-disable-next-line divide-before-multiply
+        uint256 estimationRequiredMetapoolLpTokens = curvePool
+            .get_virtual_price()
+            .divPrecisely(metapool.get_virtual_price())
+            .mulTruncate(num3CrvTokens);
+
+        int128 metapool3CrvCoinIndex = int128(
+            _getMetapoolCoinIndex(address(pTokenAddress))
+        );
+
+        // add 10% margin to the calculation of required tokens
+        // slither-disable-next-line divide-before-multiply
+        uint256 estimatedMetapoolLPWithMargin = (estimationRequiredMetapoolLpTokens *
+                1100) / 1e3;
+        uint256 crv3ReceivedWithMargin = metapool.calc_withdraw_one_coin(
+            estimatedMetapoolLPWithMargin,
+            metapool3CrvCoinIndex
+        );
+        uint256 requiredMetapoolLpTokens = (estimatedMetapoolLPWithMargin *
+            num3CrvTokens) / crv3ReceivedWithMargin;
+
+        require(
+            requiredMetapoolLpTokens <= gaugeTokens,
+            string(
+                bytes.concat(
+                    bytes("Attempting to withdraw "),
+                    bytes(Strings.toString(requiredMetapoolLpTokens)),
+                    bytes(", metapoolLP but only "),
+                    bytes(Strings.toString(gaugeTokens)),
+                    bytes(" available.")
+                )
+            )
+        );
+
+        // withdraw and unwrap with claim takes back the lpTokens and also collects the rewards for deposit
+        IRewardStaking(cvxRewardStakerAddress).withdrawAndUnwrap(
+            requiredMetapoolLpTokens,
+            true
+        );
+
+        // slither-disable-next-line unused-return
+        metapool.remove_liquidity_one_coin(
+            metapoolLPToken.balanceOf(address(this)),
+            metapool3CrvCoinIndex,
+            num3CrvTokens
+        );
+    }
+
+    function _lpWithdrawAll() internal override {
+        uint256 gaugeTokens = IRewardStaking(cvxRewardStakerAddress).balanceOf(
+            address(this)
+        );
+        IRewardStaking(cvxRewardStakerAddress).withdrawAndUnwrap(
+            gaugeTokens,
+            true
+        );
+
+        uint128 metapool3CrvCoinIndex = _getMetapoolCoinIndex(
+            address(pTokenAddress)
+        );
+        // always withdraw all of the available metapool LP tokens (similar to how we always deposit all)
+        // slither-disable-next-line unused-return
+        metapool.remove_liquidity_one_coin(
```

Comment:
This WithdrawAll method is super dangerous. There's no check on the minimum amount returned, which means that a MEV sandwich on a low liquidly pool could lose most of our funds, unless we wrapped the transaction in something that checks the funds after the withdrawAll(). 

Note that our existing curve / convex strategies don't have this same risk on withdrawAll(), because everything they are withdrawing is a mix of stablecoins that we can return to users. If an attacker bends the pool, we might not get the coins we were expecting, but we would be getting more coins total and a positive rebase. But here, we could purely lose big money.

A further danger is that this kind of MEV sandwich would be totally automated, and not even require anyone to know about this contract.

Even without a MEV attack, we could take arbitrary slippage losses from low pool liquidity with this method. If we ended up having  2/3 of a pools liquidity, us leaving would be _very_ painful here.

In the past, we've regarded withdrawAll() on our strategies as a reasonably safe step to take take in an emergency. Now here it's the most dangerous method on any strategy.

(I think all the above may apply to the regular _lpWithdraw method as well.)

### Options for fixing it

1. We revert if there is more than 1% slippage on deposits, however this is unacceptable on withdraws I think because if something depegs to 97%, we may need to just take the loss and get out.

2. Pass in a required minimum swap result amounts. Maybe two of them. This is might the best long term solution. However, our vault and our vault to strategy API's don't currently support this. We would need to alter all strategies.

3. Add code to the vault withdraw and reallocate methods to total the Vault before and after the withdraw, to check against a passed in minimum loss amount. This has the advantage of not altering the vault to strategy API's, and the disadvantage of adding 400K gas to each inner strategist action that touches withdraws. Still, the only big disadvantage is gas.

4. Up our ops game another level on strategist actions and emergency response. Make sure we always run these methods wrapped with checkLosses. (We have always been doing this for curve related transactions, this just ups the stakes a lot on getting it wrong)

Open for other thoughts too.

Reply:
Oh my thanks for catching this, and yes you make a good point. Thinking about the options you lay out: 

1. As you say this one doesn't cover all the use cases
2. This is one thing that would definitely work, though with the downside you point out, that the API changes and we need to re-deploy almost everything. 
3. I like this one a a lot. 
4. Would work though might be risky, especially since is something we need to manually keep in mind. 

I am leaning towards option 3. Though I have another proposal: 

5. Have a maxWithdrawalSlippage that is set to 1% at initialisation. And that rate is calculated when withdrawing liquidity. Also have a public setter, so that amount can be changed. This way no API would require changing and not much additional gas should be used. (too bad we forgot to add a `int256[99] private _reserved;` to BaseCurveStrategy, since it would fit nicely in there.